### PR TITLE
[SPARK-30638][CORE] Add resources allocated to PluginContext

### PR DESCRIPTION
**Pull request by tgravescs**
### What changes were proposed in this pull request?

Add the allocated resources to parameters to the PluginContext so that any plugins in driver or executor could use this information to initialize devices or use this information in a useful manner.

### Why are the changes needed?

To allow users to initialize/track devices once at the executor level before each task runs to use them.

### Does this PR introduce any user-facing change?

Yes to the people using the Executor/Driver plugin interface.

### How was this patch tested?

Unit tests and manually by writing a plugin that initialized GPU's using this interface.

### CODE REVIEW

**Comment by tgravescs** [comment]
@vanzin 
this may conflict with https://github.com/apache/spark/pull/26670

**Comment by vanzin** [comment]
I'll try to resolve conflicts if changes cause them, but I'll have very limited bandwidth...

File: core/src/test/scala/org/apache/spark/internal/plugin/PluginContainerSuite.scala [hunk 1]
```diff
@@ -140,34 +150,134 @@ class PluginContainerSuite extends SparkFunSuite with BeforeAndAfterEach with Lo
...
   }
 }
 
 object NonLocalModeSparkPlugin {
   val TEST_PATH_CONF = "spark.nonLocalPlugin.path"
+  var driverContext: PluginContext = _
+  val executorFileStr = "EXECUTOR_FILE_"
+  val driverFileStr = "DRIVER_FILE_"
+
+
```
Comment (revans2) [inline core/src/test/scala/org/apache/spark/internal/plugin/PluginContainerSuite.scala]: nit: extra line.

**Review by revans2 — APPROVED** [review]
I just had one minor nit.  I read through the code and traced through the paths.  It looks good to me.  Just sending the resources to each of the executors as they start up, and in turn to each of the plugins.

**Comment by tgravescs** [comment]
Sorry for the late change, but since the code freeze for 3.0 is this friday and it doesn't look like the stage level scheduling is going to get fully in, I'm actually going to back out the ResourceProfile addition to the api.  We can add it back in once those classes are fully public.

**Comment by tgravescs** [comment]
test failure unrelated

**Comment by tgravescs** [comment]
test this please

**Comment by tgravescs** [comment]
@vanzin if you have time to review - or no objections to api addition?

**Review by revans2 — APPROVED** [review]
+1

File: core/src/main/scala/org/apache/spark/resource/ResourceProfileBuilder.scala [hunk 2]
```diff
@@ -31,7 +31,7 @@ import org.apache.spark.annotation.Evolving
  * requirements between stages.
  */
 @Evolving
-class ResourceProfileBuilder() {
+private [spark] class ResourceProfileBuilder() {
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/resource/ResourceProfileBuilder.scala]: Shall we remove `@Evolving` if we mark this as `private [spark]`?

Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/resource/ResourceProfileBuilder.scala]: BTW, is there any reason to make this `private` in this PR?

**Comment by vanzin** [comment]
> @vanzin if you have time to review - or no objections to api addition?

I took a very quick looks and looks ok. Adding stuff go `PluginContext` should be always safe.

File: core/src/main/scala/org/apache/spark/internal/plugin/PluginContainer.scala [hunk 3]
```diff
@@ -133,17 +140,27 @@ object PluginContainer {
 
   val EXTRA_CONF_PREFIX = "spark.plugins.internal.conf."
 
-  def apply(sc: SparkContext): Option[PluginContainer] = PluginContainer(Left(sc))
+  def apply(
+      sc: SparkContext,
+      resources: java.util.Map[String, ResourceInformation]
+  ): Option[PluginContainer] = PluginContainer(Left(sc), resources)
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/internal/plugin/PluginContainer.scala]: indentation?

File: core/src/main/scala/org/apache/spark/internal/plugin/PluginContainer.scala [hunk 4]
```diff
@@ -133,17 +140,27 @@ object PluginContainer {
...
 
-  def apply(env: SparkEnv): Option[PluginContainer] = PluginContainer(Right(env))
+  def apply(
+      env: SparkEnv,
+      resources: java.util.Map[String, ResourceInformation]
+  ): Option[PluginContainer] = PluginContainer(Right(env), resources)
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/internal/plugin/PluginContainer.scala]: indentation?

File: core/src/main/scala/org/apache/spark/internal/plugin/PluginContainer.scala [hunk 5]
```diff
@@ -133,17 +140,27 @@ object PluginContainer {
...
 
-  private def apply(ctx: Either[SparkContext, SparkEnv]): Option[PluginContainer] = {
+  private def apply(
+      ctx: Either[SparkContext, SparkEnv],
+      resources: java.util.Map[String, ResourceInformation]
+  ): Option[PluginContainer] = {
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/internal/plugin/PluginContainer.scala]: ditto.

File: core/src/main/scala/org/apache/spark/internal/plugin/PluginContainer.scala [hunk 6]
```diff
@@ -133,17 +140,27 @@ object PluginContainer {
...
     val conf = ctx.fold(_.conf, _.conf)
     val plugins = Utils.loadExtensions(classOf[SparkPlugin], conf.get(PLUGINS).distinct, conf)
     if (plugins.nonEmpty) {
       ctx match {
-        case Left(sc) => Some(new DriverPluginContainer(sc, plugins))
-        case Right(env) => Some(new ExecutorPluginContainer(env, plugins))
+        case Left(sc) => Some(new DriverPluginContainer(sc, resources, plugins))
+        case Right(env) =>
+          Some(new ExecutorPluginContainer(env, resources, plugins))
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/internal/plugin/PluginContainer.scala]: Shall we make this one-liner?
```scala
case Right(env) => Some(new ExecutorPluginContainer(env, resources, plugins))
```

File: core/src/main/scala/org/apache/spark/resource/ResourceProfile.scala [hunk 7]
```diff
@@ -36,7 +36,7 @@ import org.apache.spark.internal.config.Python.PYSPARK_EXECUTOR_MEMORY
  * This is meant to be immutable so user can't change it after building.
  */
 @Evolving
-class ResourceProfile(
+private[spark] class ResourceProfile(
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/resource/ResourceProfile.scala]: Let's remove `@Evolving` for `private[spark]`.

File: core/src/test/scala/org/apache/spark/internal/plugin/PluginContainerSuite.scala [hunk 8]
```diff
@@ -140,34 +147,124 @@ class PluginContainerSuite extends SparkFunSuite with BeforeAndAfterEach with Lo
...
+  val executorFileStr = "EXECUTOR_FILE_"
+  val driverFileStr = "DRIVER_FILE_"
+
+  private def createFileStringWithGpuAddrs(
+      id: String,
+      resources: Map[String, ResourceInformation]): String = {
+    // try to keep this simple and only write the gpus addresses, if we add more resources need to
+    // make more complex
+    val resourcesString = resources.filterKeys(_.equals(GPU)).map {
+      case (name, ri) =>
```
Comment (dongjoon-hyun) [inline core/src/test/scala/org/apache/spark/internal/plugin/PluginContainerSuite.scala]: nit.
```scala
- case (name, ri) =>
+ case (_, ri) =>
```

File: core/src/main/scala/org/apache/spark/resource/ResourceProfileBuilder.scala (hunk 2 above)
Comment (tgravescs) [reply core/src/main/scala/org/apache/spark/resource/ResourceProfileBuilder.scala]: the intention was to be private until rest of stage level scheduling done and just remove private, so would rather not remove evolving, but you are right this really should be separated out as its not necessarily related, I just noticed it missing, I can separate it out.

File: core/src/main/scala/org/apache/spark/internal/plugin/PluginContainer.scala (hunk 3 above)
Comment (tgravescs) [reply core/src/main/scala/org/apache/spark/internal/plugin/PluginContainer.scala]: hmm, indentation is correct by IntelliJ scalastyle, I've seen it both ways, I'll change it though as I think this is less used.

File: core/src/main/scala/org/apache/spark/resource/ResourceProfile.scala (hunk 7 above)
Comment (tgravescs) [reply core/src/main/scala/org/apache/spark/resource/ResourceProfile.scala]: same as above

File: core/src/test/scala/org/apache/spark/internal/plugin/PluginContainerSuite.scala [hunk 9]
```diff
@@ -140,34 +147,124 @@ class PluginContainerSuite extends SparkFunSuite with BeforeAndAfterEach with Lo
...
+      eventually(timeout(10.seconds), interval(100.millis)) {
+        children = dir.listFiles()
+        assert(children != null)
+        // we have 2 discovery scripts and then expect 1 driver and 1 executor file
+        assert(children.length >= 4)
+      }
+      val execFiles =
+        children.filter(_.getName.startsWith(NonLocalModeSparkPlugin.executorFileStr))
+      assert(execFiles.size === 1)
+      val allLines = Files.readLines(execFiles(0), StandardCharsets.US_ASCII)
```
Comment (dongjoon-hyun) [inline core/src/test/scala/org/apache/spark/internal/plugin/PluginContainerSuite.scala]: `StandardCharsets.US_ASCII` -> `StandardCharsets.UTF_8`?

File: core/src/test/scala/org/apache/spark/internal/plugin/PluginContainerSuite.scala [hunk 10]
```diff
@@ -140,34 +147,124 @@ class PluginContainerSuite extends SparkFunSuite with BeforeAndAfterEach with Lo
...
       }
     }
   }
 
   override def executorPlugin(): ExecutorPlugin = {
     new ExecutorPlugin() {
       override def init(ctx: PluginContext, extraConf: JMap[String, String]): Unit = {
-        NonLocalModeSparkPlugin.writeFile(ctx.conf(), ctx.executorID())
+        NonLocalModeSparkPlugin.writeFile(NonLocalModeSparkPlugin.executorFileStr, ctx.conf(),
+        ctx.executorID(), ctx.resources().asScala.toMap)
```
Comment (dongjoon-hyun) [inline core/src/test/scala/org/apache/spark/internal/plugin/PluginContainerSuite.scala]: indentation?

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. (with minor comments)
I believe my comments can be addressed later.

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by tgravescs** [comment]
thanks @dongjoon-hyun I'll file a followup to fix the couple nits left.

**Pull request closed without merging**